REMOVE_SPENT_UTXOS=true causes /network/status to fail with "Unable to rollback against JDBC Connection"

# `REMOVE_SPENT_UTXOS=true` Failure Flow

Date: 2026-04-09

This note captures the live failure path reproduced on the current mainnet Helm deployment.

## Reproduced symptom

A valid in-pod request to `POST /network/status` returned:

```json
{
  "code": 5000,
  "message": "An error occurred",
  "retriable": true,
  "details": {
    "message": "An error occurred for request 17: Unable to rollback against JDBC Connection"
  }
}
```

The same request path does not fail when `REMOVE_SPENT_UTXOS=false`.

## High-level conclusion

The visible rollback error is a symptom, not the original cause.

The actual chain is:

1. `/network/status` tries to fetch discovered peers from yaci-indexer.
2. That HTTP call fails while yaci is not ready.
3. The API incorrectly sets the current request thread to interrupted state.
4. With pruning enabled, the same interrupted request thread then runs the extra `oldest_block_identifier` DB query.
5. PostgreSQL JDBC sees that interrupt while doing socket I/O, the socket is closed, and the connection is marked broken.
6. Transaction rollback then fails because the JDBC connection is already closed.

## Flow diagram

```mermaid
flowchart TD
    A[POST /network/status] --> B[NetworkApiImpl.networkStatus]
    B --> C[NetworkServiceImpl.getNetworkStatus]

    C --> D[findLatestBlockIdentifier]
    D --> D1[OK]

    C --> E[findGenesisBlockIdentifier]
    E --> E1[OK]

    C --> F[topologyConfigService.getPeers]
    F --> G[yaciHttpGateway.getDiscoveredPeers]
    G --> H[yaci HTTP request fails]
    H --> I["catch (IOException | InterruptedException)"]
    I --> J["Thread.currentThread.interrupt marks current request thread interrupted"]
    J --> K[getPeers falls back to static peers]

    K --> L{REMOVE_SPENT_UTXOS}
    L -->|false| M[Skip oldest_block_identifier]
    M --> N[Request can complete]

    L -->|true| O["findOldestBlockIdentifier on same interrupted request thread"]
    J -. "interrupt flag still set" .-> O
    O --> P[blockRepository.findBlockProjectionByNumber]
    P --> Q["PostgreSQL JDBC does socket I/O on interrupted thread"]
    Q --> R[SocketException: Closed by interrupt]
    R --> S[PSQLException SQLSTATE 08006]
    S --> T[Hikari marks connection broken]
    T --> U[Rollback hits closed connection]
    U --> V["500 Unable to rollback against JDBC Connection"]
```

## Live evidence

- The API request entered `NetworkServiceImpl.networkStatus(...)`.
- `TopologyConfigServiceImpl.getPeers()` called yaci peer discovery first.
- `YaciHttpGatewayImpl` logged `Error during yaci-indexer peers HTTP request`.
- The same request later failed inside `LedgerBlockServiceImpl.findOldestBlockIdentifier(...)`.
- Hikari logged `SQLSTATE(08006)`.
- PostgreSQL JDBC logged `An I/O error occurred while sending to the backend`.
- The low-level cause in the stack was `java.net.SocketException: Closed by interrupt`.
- Spring then surfaced `JpaSystemException: Unable to rollback against JDBC Connection`.

## Why pruning changes the outcome

When `REMOVE_SPENT_UTXOS=false`, `/network/status` does not execute the extra oldest-block query, so the interrupted request thread does not immediately hit that DB path.

When `REMOVE_SPENT_UTXOS=true`, `/network/status` must populate Rosetta's optional `oldest_block_identifier`, so the same interrupted request thread executes the extra DB lookup and triggers the JDBC failure.

## Important implication

This is not caused by Kubernetes probes. The current API probes use only:

- `/actuator/health/startup`
- `/actuator/health/readiness`
- `/actuator/health/liveness`

The reproduced failure came from a direct `POST /network/status` request.

## Code locations

- `api/src/main/java/org/cardanofoundation/rosetta/api/network/service/NetworkServiceImpl.java`
- `api/src/main/java/org/cardanofoundation/rosetta/api/network/service/TopologyConfigServiceImpl.java`
- `api/src/main/java/org/cardanofoundation/rosetta/client/YaciHttpGatewayImpl.java`
- `api/src/main/java/org/cardanofoundation/rosetta/api/block/service/LedgerBlockServiceImpl.java`
- `api/src/main/java/org/cardanofoundation/rosetta/api/block/model/repository/BlockRepository.java`

## Most likely bug

The strongest current explanation is an interruption-handling bug in `YaciHttpGatewayImpl`:

- it catches `IOException | InterruptedException`
- it always calls `Thread.currentThread().interrupt()`
- that leaves the request thread interrupted even for ordinary I/O failures
- the next JDBC call in the same request then fails with `Closed by interrupt`

This should be treated as the primary suspect for the current `REMOVE_SPENT_UTXOS=true` failure path.

## Captured stack traces

These are the distinct error stacks captured from the live `rosetta-api` pod for the reproduced `POST /network/status` request on `2026-04-09T10:41:39Z`.

The repeated Spring Security / Undertow servlet filter frames are trimmed where they do not change the exception chain. The application frames and cause chains are preserved.

### 1. yaci peer discovery HTTP failure

<details>
<summary>Open yaci peer discovery HTTP failure stack trace</summary>

```text
2026-04-09T10:41:39.824Z ERROR 1 --- [    undertow-16] o.c.rosetta.client.YaciHttpGatewayImpl   : Error during yaci-indexer peers HTTP request

java.net.ConnectException: null
	at java.net.http/jdk.internal.net.http.HttpClientImpl.send(HttpClientImpl.java:923) ~[java.net.http:na]
	at java.net.http/jdk.internal.net.http.HttpClientFacade.send(HttpClientFacade.java:133) ~[java.net.http:na]
	at org.cardanofoundation.rosetta.client.YaciHttpGatewayImpl.getDiscoveredPeers(YaciHttpGatewayImpl.java:87) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.service.TopologyConfigServiceImpl.getDiscoveredPeers(TopologyConfigServiceImpl.java:107) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.service.TopologyConfigServiceImpl.getPeers(TopologyConfigServiceImpl.java:45) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.service.NetworkServiceImpl.networkStatus(NetworkServiceImpl.java:155) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.service.NetworkServiceImpl.getNetworkStatus(NetworkServiceImpl.java:125) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.controller.NetworkApiImpl.networkStatus(NetworkApiImpl.java:50) ~[!/:na]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:565) ~[na:na]
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:359) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:196) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.validation.beanvalidation.MethodValidationInterceptor.invoke(MethodValidationInterceptor.java:174) ~[spring-context-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:728) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.cardanofoundation.rosetta.api.network.controller.NetworkApiImpl$$SpringCGLIB$$0.networkStatus(<generated>) ~[!/:na]
	...
Caused by: java.net.ConnectException: null
	at java.net.http/jdk.internal.net.http.common.Utils.toConnectException(Utils.java:1046) ~[java.net.http:na]
	at java.net.http/jdk.internal.net.http.PlainHttpConnection.connectAsync(PlainHttpConnection.java:215) ~[java.net.http:na]
	at java.net.http/jdk.internal.net.http.PlainHttpConnection.checkRetryConnect(PlainHttpConnection.java:268) ~[java.net.http:na]
	at java.net.http/jdk.internal.net.http.PlainHttpConnection.lambda$connectAsync$0(PlainHttpConnection.java:226) ~[java.net.http:na]
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:980) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:957) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:556) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1819) ~[na:na]
	at java.base/java.util.concurrent.ThreadPerTaskExecutor$TaskRunner.run(ThreadPerTaskExecutor.java:291) ~[na:na]
	... 1 common frames omitted
Caused by: java.nio.channels.ClosedChannelException: null
	at java.base/sun.nio.ch.SocketChannelImpl.ensureOpen(SocketChannelImpl.java:204) ~[na:na]
	at java.base/sun.nio.ch.SocketChannelImpl.connect(SocketChannelImpl.java:942) ~[na:na]
	at java.net.http/jdk.internal.net.http.PlainHttpConnection.connectAsync(PlainHttpConnection.java:203) ~[java.net.http:na]
	... 8 common frames omitted
```

</details>

### 2. Fallback `ApiException` after peer discovery failed

<details>
<summary>Open fallback ApiException stack trace</summary>

```text
2026-04-09T10:41:39.827Z  WARN 1 --- [    undertow-16] o.c.r.a.n.s.TopologyConfigServiceImpl    : [getDiscoveredPeers] Failed to fetch discovered peers!

org.cardanofoundation.rosetta.common.exception.ApiException: Unable to get data from the downstream gateway
	at org.cardanofoundation.rosetta.common.exception.ExceptionFactory.gatewayError(ExceptionFactory.java:301) ~[!/:na]
	at org.cardanofoundation.rosetta.client.YaciHttpGatewayImpl.getDiscoveredPeers(YaciHttpGatewayImpl.java:108) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.service.TopologyConfigServiceImpl.getDiscoveredPeers(TopologyConfigServiceImpl.java:107) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.service.TopologyConfigServiceImpl.getPeers(TopologyConfigServiceImpl.java:45) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.service.NetworkServiceImpl.networkStatus(NetworkServiceImpl.java:155) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.service.NetworkServiceImpl.getNetworkStatus(NetworkServiceImpl.java:125) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.controller.NetworkApiImpl.networkStatus(NetworkApiImpl.java:50) ~[!/:na]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:565) ~[na:na]
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:359) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:196) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.validation.beanvalidation.MethodValidationInterceptor.invoke(MethodValidationInterceptor.java:174) ~[spring-context-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:728) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.cardanofoundation.rosetta.api.network.controller.NetworkApiImpl$$SpringCGLIB$$0.networkStatus(<generated>) ~[!/:na]
	...
```

</details>

### 3. JDBC connection broken during `findOldestBlockIdentifier(...)`

<details>
<summary>Open JDBC SQLSTATE 08006 / Closed by interrupt stack trace</summary>

```text
2026-04-09T10:41:39.839Z  WARN 1 --- [    undertow-16] com.zaxxer.hikari.pool.ProxyConnection   : CardanoRosettaJavaDBPool - Connection org.postgresql.jdbc.PgConnection@5906ebfb marked as broken because of SQLSTATE(08006), ErrorCode(0)

org.postgresql.util.PSQLException: An I/O error occurred while sending to the backend.
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:399) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:517) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:434) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:194) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:137) ~[postgresql-42.7.4.jar!/:42.7.4]
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeQuery(ProxyPreparedStatement.java:52) ~[HikariCP-6.3.0.jar!/:na]
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeQuery(HikariProxyPreparedStatement.java) ~[HikariCP-6.3.0.jar!/:na]
	at org.hibernate.sql.results.jdbc.internal.DeferredResultSetAccess.executeQuery(DeferredResultSetAccess.java:250) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.sql.results.jdbc.internal.DeferredResultSetAccess.getResultSet(DeferredResultSetAccess.java:171) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.sql.results.jdbc.internal.JdbcValuesResultSetImpl.<init>(JdbcValuesResultSetImpl.java:74) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.resolveJdbcValuesSource(JdbcSelectExecutorStandardImpl.java:355) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.doExecuteQuery(JdbcSelectExecutorStandardImpl.java:137) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.executeQuery(JdbcSelectExecutorStandardImpl.java:102) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.sql.exec.spi.JdbcSelectExecutor.executeQuery(JdbcSelectExecutor.java:91) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.sql.exec.spi.JdbcSelectExecutor.list(JdbcSelectExecutor.java:165) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.lambda$new$1(ConcreteSqmSelectQueryPlan.java:152) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.withCacheableSqmInterpretation(ConcreteSqmSelectQueryPlan.java:442) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.performList(ConcreteSqmSelectQueryPlan.java:362) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.query.sqm.internal.QuerySqmImpl.doList(QuerySqmImpl.java:380) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.query.spi.AbstractSelectionQuery.list(AbstractSelectionQuery.java:143) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.query.spi.AbstractSelectionQuery.getSingleResult(AbstractSelectionQuery.java:275) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.springframework.data.jpa.repository.query.JpaQueryExecution$SingleEntityExecution.doExecute(JpaQueryExecution.java:224) ~[spring-data-jpa-3.5.0.jar!/:3.5.0]
	at org.springframework.data.jpa.repository.query.JpaQueryExecution.execute(JpaQueryExecution.java:93) ~[spring-data-jpa-3.5.0.jar!/:3.5.0]
	at org.springframework.data.jpa.repository.query.AbstractJpaQuery.doExecute(AbstractJpaQuery.java:159) ~[spring-data-jpa-3.5.0.jar!/:3.5.0]
	at org.springframework.data.jpa.repository.query.AbstractJpaQuery.execute(AbstractJpaQuery.java:147) ~[spring-data-jpa-3.5.0.jar!/:3.5.0]
	at org.springframework.data.repository.core.support.RepositoryMethodInvoker.doInvoke(RepositoryMethodInvoker.java:170) ~[spring-data-commons-3.5.0.jar!/:3.5.0]
	at org.springframework.data.repository.core.support.RepositoryMethodInvoker.invoke(RepositoryMethodInvoker.java:158) ~[spring-data-commons-3.5.0.jar!/:3.5.0]
	at org.springframework.data.repository.core.support.QueryExecutorMethodInterceptor.doInvoke(QueryExecutorMethodInterceptor.java:170) ~[spring-data-commons-3.5.0.jar!/:3.5.0]
	at org.springframework.data.repository.core.support.QueryExecutorMethodInterceptor.invoke(QueryExecutorMethodInterceptor.java:149) ~[spring-data-commons-3.5.0.jar!/:3.5.0]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:69) ~[spring-data-commons-3.5.0.jar!/:3.5.0]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:380) ~[spring-tx-6.2.7.jar!/:6.2.7]
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119) ~[spring-tx-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:138) ~[spring-tx-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$CrudMethodMetadataPopulatingMethodInterceptor.invoke(CrudMethodMetadataPostProcessor.java:136) ~[spring-data-jpa-3.5.0.jar!/:3.5.0]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:223) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at jdk.proxy2/jdk.proxy2.$Proxy191.findBlockProjectionByNumber(Unknown Source) ~[na:na]
	at org.cardanofoundation.rosetta.api.block.service.LedgerBlockServiceImpl.findOldestBlockIdentifier(LedgerBlockServiceImpl.java:216) ~[!/:na]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:565) ~[na:na]
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:359) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:196) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:380) ~[spring-tx-6.2.7.jar!/:6.2.7]
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119) ~[spring-tx-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:728) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.cardanofoundation.rosetta.api.block.service.LedgerBlockServiceImpl$$SpringCGLIB$$0.findOldestBlockIdentifier(<generated>) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.service.NetworkServiceImpl.networkStatus(NetworkServiceImpl.java:163) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.service.NetworkServiceImpl.getNetworkStatus(NetworkServiceImpl.java:125) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.controller.NetworkApiImpl.networkStatus(NetworkApiImpl.java:50) ~[!/:na]
	...
Caused by: java.net.SocketException: Closed by interrupt
	at java.base/java.net.Socket$SocketInputStream.implRead(Socket.java:990) ~[na:na]
	at java.base/java.net.Socket$SocketInputStream.read(Socket.java:970) ~[na:na]
	at org.postgresql.core.VisibleBufferedInputStream.readMore(VisibleBufferedInputStream.java:192) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:159) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:144) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.core.VisibleBufferedInputStream.read(VisibleBufferedInputStream.java:76) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.core.PGStream.receiveChar(PGStream.java:476) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2174) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:372) ~[postgresql-42.7.4.jar!/:42.7.4]
	... 197 common frames omitted

2026-04-09T10:41:39.841Z  WARN 1 --- [    undertow-16] o.h.engine.jdbc.spi.SqlExceptionHelper   : SQL Error: 0, SQLState: 08006
2026-04-09T10:41:39.841Z ERROR 1 --- [    undertow-16] o.h.engine.jdbc.spi.SqlExceptionHelper   : An I/O error occurred while sending to the backend.
2026-04-09T10:41:39.848Z ERROR 1 --- [    undertow-16] o.s.t.i.TransactionInterceptor           : Application exception overridden by rollback exception

org.springframework.dao.DataAccessResourceFailureException: JDBC exception executing SQL [select be1_0.hash,be1_0.body_size,be1_0.block_time,be1_0.epoch,be1_0.no_of_txs,be1_0.number,be1_0.prev_hash,be1_0.slot,be1_0.slot_leader from block be1_0 where be1_0.number=? order by be1_0.number fetch first 1 rows only] [An I/O error occurred while sending to the backend.] [n/a]
	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:278) ~[spring-orm-6.2.7.jar!/:6.2.7]
	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:256) ~[spring-orm-6.2.7.jar!/:6.2.7]
	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.translateExceptionIfPossible(AbstractEntityManagerFactoryBean.java:560) ~[spring-orm-6.2.7.jar!/:6.2.7]
	at org.springframework.dao.support.ChainedPersistenceExceptionTranslator.translateExceptionIfPossible(ChainedPersistenceExceptionTranslator.java:61) ~[spring-tx-6.2.7.jar!/:6.2.7]
	at org.springframework.dao.support.DataAccessUtils.translateIfNecessary(DataAccessUtils.java:343) ~[spring-tx-6.2.7.jar!/:6.2.7]
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:160) ~[spring-tx-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$CrudMethodMetadataPopulatingMethodInterceptor.invoke(CrudMethodMetadataPostProcessor.java:136) ~[spring-data-jpa-3.5.0.jar!/:3.5.0]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:223) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at jdk.proxy2/jdk.proxy2.$Proxy191.findBlockProjectionByNumber(Unknown Source) ~[na:na]
	at org.cardanofoundation.rosetta.api.block.service.LedgerBlockServiceImpl.findOldestBlockIdentifier(LedgerBlockServiceImpl.java:216) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.service.NetworkServiceImpl.networkStatus(NetworkServiceImpl.java:163) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.service.NetworkServiceImpl.getNetworkStatus(NetworkServiceImpl.java:125) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.controller.NetworkApiImpl.networkStatus(NetworkApiImpl.java:50) ~[!/:na]
	...
Caused by: org.hibernate.exception.JDBCConnectionException: JDBC exception executing SQL [select be1_0.hash,be1_0.body_size,be1_0.block_time,be1_0.epoch,be1_0.no_of_txs,be1_0.number,be1_0.prev_hash,be1_0.slot,be1_0.slot_leader from block be1_0 where be1_0.number=? order by be1_0.number fetch first 1 rows only] [An I/O error occurred while sending to the backend.] [n/a]
	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:100) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:58) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:108) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:94) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.sql.results.jdbc.internal.DeferredResultSetAccess.executeQuery(DeferredResultSetAccess.java:268) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.sql.results.jdbc.internal.DeferredResultSetAccess.getResultSet(DeferredResultSetAccess.java:171) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.sql.results.jdbc.internal.JdbcValuesResultSetImpl.<init>(JdbcValuesResultSetImpl.java:74) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.resolveJdbcValuesSource(JdbcSelectExecutorStandardImpl.java:355) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.doExecuteQuery(JdbcSelectExecutorStandardImpl.java:137) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.executeQuery(JdbcSelectExecutorStandardImpl.java:102) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.sql.exec.spi.JdbcSelectExecutor.executeQuery(JdbcSelectExecutor.java:91) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.sql.exec.spi.JdbcSelectExecutor.list(JdbcSelectExecutor.java:165) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.lambda$new$1(ConcreteSqmSelectQueryPlan.java:152) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.withCacheableSqmInterpretation(ConcreteSqmSelectQueryPlan.java:442) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.performList(ConcreteSqmSelectQueryPlan.java:362) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.query.sqm.internal.QuerySqmImpl.doList(QuerySqmImpl.java:380) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.query.spi.AbstractSelectionQuery.list(AbstractSelectionQuery.java:143) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.query.spi.AbstractSelectionQuery.getSingleResult(AbstractSelectionQuery.java:275) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.springframework.data.jpa.repository.query.JpaQueryExecution$SingleEntityExecution.doExecute(JpaQueryExecution.java:224) ~[spring-data-jpa-3.5.0.jar!/:3.5.0]
	...
Caused by: org.postgresql.util.PSQLException: An I/O error occurred while sending to the backend.
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:399) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:517) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:434) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:194) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:137) ~[postgresql-42.7.4.jar!/:42.7.4]
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeQuery(ProxyPreparedStatement.java:52) ~[HikariCP-6.3.0.jar!/:na]
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeQuery(HikariProxyPreparedStatement.java) ~[HikariCP-6.3.0.jar!/:na]
	at org.hibernate.sql.results.jdbc.internal.DeferredResultSetAccess.executeQuery(DeferredResultSetAccess.java:250) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	... 190 common frames omitted
Caused by: java.net.SocketException: Closed by interrupt
	at java.base/java.net.Socket$SocketInputStream.implRead(Socket.java:990) ~[na:na]
	at java.base/java.net.Socket$SocketInputStream.read(Socket.java:970) ~[na:na]
	at org.postgresql.core.VisibleBufferedInputStream.readMore(VisibleBufferedInputStream.java:192) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:159) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:144) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.core.VisibleBufferedInputStream.read(VisibleBufferedInputStream.java:76) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.core.PGStream.receiveChar(PGStream.java:476) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2174) ~[postgresql-42.7.4.jar!/:42.7.4]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:372) ~[postgresql-42.7.4.jar!/:42.7.4]
	... 197 common frames omitted
```

</details>

### 4. Final rollback exception surfaced to the client

<details>
<summary>Open final rollback exception stack trace</summary>

```text
2026-04-09T10:41:39.852Z ERROR 1 --- [    undertow-16] o.c.r.c.e.GlobalExceptionHandler         : An error occurred during the request

org.springframework.orm.jpa.JpaSystemException: Unable to rollback against JDBC Connection
	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:348) ~[spring-orm-6.2.7.jar!/:6.2.7]
	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:256) ~[spring-orm-6.2.7.jar!/:6.2.7]
	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:241) ~[spring-orm-6.2.7.jar!/:6.2.7]
	at org.springframework.orm.jpa.JpaTransactionManager.doRollback(JpaTransactionManager.java:593) ~[spring-orm-6.2.7.jar!/:6.2.7]
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processRollback(AbstractPlatformTransactionManager.java:897) ~[spring-tx-6.2.7.jar!/:6.2.7]
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.rollback(AbstractPlatformTransactionManager.java:866) ~[spring-tx-6.2.7.jar!/:6.2.7]
	at org.springframework.transaction.interceptor.TransactionAspectSupport.completeTransactionAfterThrowing(TransactionAspectSupport.java:716) ~[spring-tx-6.2.7.jar!/:6.2.7]
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:384) ~[spring-tx-6.2.7.jar!/:6.2.7]
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119) ~[spring-tx-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:728) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.cardanofoundation.rosetta.api.block.service.LedgerBlockServiceImpl$$SpringCGLIB$$0.findOldestBlockIdentifier(<generated>) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.service.NetworkServiceImpl.networkStatus(NetworkServiceImpl.java:163) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.service.NetworkServiceImpl.getNetworkStatus(NetworkServiceImpl.java:125) ~[!/:na]
	at org.cardanofoundation.rosetta.api.network.controller.NetworkApiImpl.networkStatus(NetworkApiImpl.java:50) ~[!/:na]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:565) ~[na:na]
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:359) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:196) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.validation.beanvalidation.MethodValidationInterceptor.invoke(MethodValidationInterceptor.java:174) ~[spring-context-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:728) ~[spring-aop-6.2.7.jar!/:6.2.7]
	at org.cardanofoundation.rosetta.api.network.controller.NetworkApiImpl$$SpringCGLIB$$0.networkStatus(<generated>) ~[!/:na]
	...
Caused by: org.hibernate.TransactionException: Unable to rollback against JDBC Connection
	at org.hibernate.resource.jdbc.internal.AbstractLogicalConnectionImplementor.rollback(AbstractLogicalConnectionImplementor.java:137) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.rollback(JdbcResourceLocalTransactionCoordinatorImpl.java:289) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.hibernate.engine.transaction.internal.TransactionImpl.rollback(TransactionImpl.java:142) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	at org.springframework.orm.jpa.JpaTransactionManager.doRollback(JpaTransactionManager.java:589) ~[spring-orm-6.2.7.jar!/:6.2.7]
	... 154 common frames omitted
Caused by: java.sql.SQLException: Connection is closed
	at com.zaxxer.hikari.pool.ProxyConnection$ClosedConnection.lambda$getClosedConnection$0(ProxyConnection.java:559) ~[HikariCP-6.3.0.jar!/:na]
	at jdk.proxy3/jdk.proxy3.$Proxy120.rollback(Unknown Source) ~[na:na]
	at com.zaxxer.hikari.pool.ProxyConnection.rollback(ProxyConnection.java:386) ~[HikariCP-6.3.0.jar!/:na]
	at com.zaxxer.hikari.pool.HikariProxyConnection.rollback(HikariProxyConnection.java) ~[HikariCP-6.3.0.jar!/:na]
	at org.hibernate.resource.jdbc.internal.AbstractLogicalConnectionImplementor.rollback(AbstractLogicalConnectionImplementor.java:127) ~[hibernate-core-6.6.15.Final.jar!/:6.6.15.Final]
	... 157 common frames omitted
```

</details>

--- a/api/src/main/java/org/cardanofoundation/rosetta/client/YaciHttpGatewayImpl.java
+++ b/api/src/main/java/org/cardanofoundation/rosetta/client/YaciHttpGatewayImpl.java
@@ -65,11 +65,12 @@ public class YaciHttpGatewayImpl implements YaciHttpGateway {
             } else {
                 throw ExceptionFactory.gatewayError(false);
             }
-        } catch (IOException | InterruptedException e) {
+        } catch (InterruptedException e) {
             log.error("Error during yaci-indexer HTTP request", e);
-
             Thread.currentThread().interrupt();
-
+            throw ExceptionFactory.gatewayError(true);
+        } catch (IOException e) {
+            log.error("Error during yaci-indexer HTTP request", e);
             throw ExceptionFactory.gatewayError(true);
         }
     }
@@ -100,11 +101,12 @@ public class YaciHttpGatewayImpl implements YaciHttpGateway {
             } else {
                 throw ExceptionFactory.gatewayError(false);
             }
-        } catch (IOException | InterruptedException e) {
+        } catch (InterruptedException e) {
             log.error("Error during yaci-indexer peers HTTP request", e);
-
             Thread.currentThread().interrupt();
-
+            throw ExceptionFactory.gatewayError(true);
+        } catch (IOException e) {
+            log.error("Error during yaci-indexer peers HTTP request", e);
             throw ExceptionFactory.gatewayError(true);
         }
     }

